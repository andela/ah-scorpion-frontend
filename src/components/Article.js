import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import React from 'react';

const getReadingTime = (body) => {
  const blocks = JSON.parse(body).blocks;
  const text = blocks.map(obj => obj.text);
  const words = text.join(' ');
  const wordsPerMinute = 270;
  const time = Math.round(words.split(' ').length / wordsPerMinute);
  const readTime = time <= 0 ? 1 : time;
  return readTime;
};

const Article = ({
  title,
  description,
  image,
  createdAt,
  slug,
  averageRating,
  likes,
  dislikes,
  author,
  body,
}) => (
  <div className="col-4 p-5">
    <div className="card img-card pb-2">
      <h5 className="card-title text-center pt-2">
        <Link to={`/article/${slug}`} className="article-title">
          {title}
        </Link>
      </h5>
      <img
        className="card-img-top card-image-custom"
        alt=""
        style={{ backgroundColor: 'rgb(21, 127, 251)' }}
        src={image}
      />
      <div className="card-body">
        <p className="card-text card-overflow">{description}</p>
        <br />
        <small className="text-muted">
          author:
          {` ${author}`}
        </small>
        <div className="d-flex justify-content-between align-items-center">
          <small className="text-muted">
            created:
            {` ${new Date(createdAt).toDateString()}`}
          </small>
        </div>
        <small className="text-muted">{getReadingTime(body)} min read</small>
      </div>
      <div className="row px-4">
        <div className="col-1 text-center">
          <i className="text-warning fa fa-star" />
          <br />
          <small>{averageRating || ' - '}</small>
        </div>
        <div className="col-1 text-center">
          <i className="text-success fa fa-thumbs-up" />
          <br />
          <small>{likes}</small>
        </div>
        <div className="col-1 text-center">
          <i className="text-danger fa fa-thumbs-down" />
          <br />
          <small>{dislikes}</small>
        </div>
        <div className="ml-auto">
          <Link to={`/article/${slug}`} className="ml-auto">
            <button type="button" className="btn btn-sm btn-outline-primary">
              Read More
            </button>
          </Link>
        </div>
      </div>
    </div>
  </div>
);

Article.propTypes = {
  title: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  createdAt: PropTypes.string.isRequired,
  slug: PropTypes.string.isRequired,
  averageRating: PropTypes.number,
  likes: PropTypes.number.isRequired,
  dislikes: PropTypes.number.isRequired,
  image: PropTypes.string.isRequired,
  author: PropTypes.string.isRequired,
  body: PropTypes.shape().isRequired,
};

Article.defaultProps = {
  averageRating: null,
};

export default Article;
