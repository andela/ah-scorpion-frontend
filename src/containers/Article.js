import React, { Component } from 'react';
import Footer from '../components/Footer';
import ArticleExample from '../components/ArticleExample';

class Article extends Component {
  render() {
    return (
      <div>
        <ArticleExample />
        <Footer />
      </div>
    );
  }
}

export default Article;
