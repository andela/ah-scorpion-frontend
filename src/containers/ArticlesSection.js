import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import axios from 'axios';
import ReactPaginate from 'react-paginate';
import Footer from '../components/Footer';
import handleGetArticles from '../actions/getArticles';
import Loader from '../components/Loader';
import ArticlesList from '../components/ArticleList';
import './MyArticlesPage.css';

const { REACT_APP_BASE_URL } = process.env;
const articlesUrl = `${REACT_APP_BASE_URL}/articles/`;


class ArticlesSection extends Component {
  constructor(props) {
    super(props);
    this.state = {
      offset: 0,
      perPage: 6,
      pageCount: 1,
    };
  }

  componentWillMount() {
    const { getArticles } = this.props;
    const { perPage, offset } = this.state;
    getArticles(perPage, offset);
    axios.get(articlesUrl).then(({ data }) => {
      const pageCount = Math.ceil(data.length / perPage);
      this.setState({ pageCount });
    }).catch((error) => {
      console.log(error);
    });
  }

  handlePageClick = (data) => {
    const { selected } = data;
    const { perPage } = this.state;
    const { getArticles } = this.props;
    const offset = Math.ceil(selected * perPage);
    this.setState({ offset }, () => {
      getArticles(perPage, offset);
    });
  };


  render() {
    const {
      isFetching, articles, fetchFailure, errorMessage,
    } = this.props;
    const { pageCount } = this.state;

    return (
      <React.Fragment>
        <main>
          <div className="text-center py-5 dashboard-banner" style={{ height: '350px' }}>
            <div className="container">
              <div className="row my-5 justify-content-center">
                <div className="col-md-9">
                  <h1>Interesting Reads</h1>
                  <p className="lead text-white">
                    The articles are expertly written and captivating, the creators Make them short
                    and sweet, but not too short so folks don
                    {"'"}
t simply skip over them entirely.
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div className="py-4 bg-light">
            <div className="container">
              {isFetching ? (
                <div className="text-center">
                  <Loader />
                </div>
              ) : (
                <div className="row">
                  <ArticlesList
                    articles={articles}
                    fetchFailure={fetchFailure}
                    errorMessage={errorMessage}
                  />
                </div>
              )}
              <ReactPaginate
                previousLabel="previous"
                nextLabel="next"
                breakLabel={<a href="">...</a>}
                breakClassName="break-me"
                pageCount={pageCount}
                marginPagesDisplayed={2}
                pageRangeDisplayed={5}
                onPageChange={this.handlePageClick}
                containerClassName="pagination"
                activeClassName="page-item active"
                disabledClassName="page-item disabled"
                pageClassName="page-item"
                extraAriaContext="Page navigation example"
              />
            </div>
          </div>
        </main>
        <Footer />
      </React.Fragment>
    );
  }
}

ArticlesSection.propTypes = {
  isFetching: PropTypes.bool.isRequired,
  articles: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  fetchFailure: PropTypes.bool.isRequired,
  errorMessage: PropTypes.string.isRequired,
  getArticles: PropTypes.func.isRequired,
};

const mapDispatchToProps = dispatch => ({
  getArticles: (limit, offset) => dispatch(handleGetArticles(limit, offset)),
});

const mapStateToProps = ({ AllArticles }) => AllArticles;

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ArticlesSection);
