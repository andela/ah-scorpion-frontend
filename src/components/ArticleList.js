import PropTypes from 'prop-types';
import React from 'react';
import Article from './Article';

const ArticlesList = ({ articles, fetchFailure, errorMessage }) => {
  if (fetchFailure) {
    return <div className="alert alert-danger text-center">{errorMessage}</div>;
  }

  if (articles.length === 0) {
    return <h4>There is currently no article available.</h4>;
  }

  return articles.map((article) => {
    const imgBase = JSON.parse(article.body).entityMap[0];
    return (
      <Article
        title={article.title}
        createdAt={article.createdAt}
        description={article.description}
        slug={article.slug}
        key={article.slug}
        averageRating={article.averageRating}
        likes={article.likes}
        dislikes={article.dislikes}
        image={imgBase ? imgBase.data.src : ''}
        author={article.author.username}
      />
    );
  });
};

ArticlesList.propTypes = {
  articles: PropTypes.arrayOf(PropTypes.shape()).isRequired,
};

export default ArticlesList;
