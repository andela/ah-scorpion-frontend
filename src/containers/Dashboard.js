import React, { Component } from 'react';

import PropTypes from 'prop-types';
import UserNavBar from '../components/UserNavBar';
import Footer from '../components/Footer';
import ArticlesSection from './ArticlesSection';

class Dashboard extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  render() {
    const { history } = this.props;
    return (
      <div>
        <UserNavBar history={history} />
        <ArticlesSection />
        <Footer />
      </div>
    );
  }
}

Dashboard.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
};

export default Dashboard;
