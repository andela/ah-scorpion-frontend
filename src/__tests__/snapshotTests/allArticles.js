import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { shallow } from 'enzyme';
import shallowToJSON from 'enzyme-to-json';
import Article from '../../components/Article';
import ArticleList from '../../components/ArticleList';
import ArticlesSection from '../../containers/ArticlesSection';
import Dashboard from '../../containers/Dashboard';

describe('The Article component', () => {
  it('Should not regress', () => {
    const tree = shallow(
      <MemoryRouter initialEntries={[{ pathname: '/', key: 'testKey' }]}>
        <Article
          title="test title"
          description="test description"
          image=""
          createdAt="2018-09-12T18:51:24.099861Z"
          slug="test-slug-001"
          averageRating={1}
          likes={0}
          dislikes={0}
          author="lenny"
        />
      </MemoryRouter>,
    );
    expect(shallowToJSON(tree)).toMatchSnapshot();
  });
});

describe('The ArticleList component', () => {
  const testProps = {
    articles: [],
    fetchFailure: false,
    errorMessage: '',
  };
  const articles = [
    {
      id: 43,
      title: "Lenny's First Article 1",
      author: {
        email: 'lennykmutua@gmail.com',
        username: 'lenny',
        bio: '',
        image: '',
      },
      likes: 0,
      dislikes: 0,
      favorited: 1,
      averageRating: null,
      ratingsCount: 0,
      slug: 'lennys-first-article-1-29b5102ef9a84eb9bfbf60da42f63122',
      body: "Like I said... I don't know what's going on. I'm just testing Hoslack's work.",
      description: "This is my first article and I don't know what's going on...",
      images: ['https://url-to-image'],
      createdAt: '2018-09-12T18:51:24.099861Z',
      updatedAt: '2018-09-12T18:51:24.100977Z',
      tagList: ['Lenny', 'First'],
    },
  ];
  it('Should render text if there are no articles', () => {
    const tree = shallow(
      <MemoryRouter initialEntries={[{ pathname: '/', key: 'testKey' }]}>
        <ArticleList {...testProps} />
      </MemoryRouter>,
    );
    expect(shallowToJSON(tree)).toMatchSnapshot();
  });
  it('Should render the error message in case of failure', () => {
    const tree = shallow(
      <MemoryRouter initialEntries={[{ pathname: '/', key: 'testKey' }]}>
        <ArticleList {...testProps} fetchFailure errorMessage="An error occurred" />
      </MemoryRouter>,
    );
    expect(shallowToJSON(tree)).toMatchSnapshot();
  });

  it('Should render all articles in case of a successful request', () => {
    const tree = shallow(
      <MemoryRouter initialEntries={[{ pathname: '/', key: 'testKey' }]}>
        <ArticleList {...testProps} articles={articles} />
      </MemoryRouter>,
    );
    expect(shallowToJSON(tree)).toMatchSnapshot();
  });
});

describe('The ArticleSection component', () => {
  const testProps = {
    articles: [],
    fetchFailure: false,
    errorMessage: '',
    isFetching: false,
  };
  const articles = [
    {
      id: 43,
      title: "Lenny's First Article 1",
      author: {
        email: 'lennykmutua@gmail.com',
        username: 'lenny',
        bio: '',
        image: '',
      },
      likes: 0,
      dislikes: 0,
      favorited: 1,
      averageRating: null,
      ratingsCount: 0,
      slug: 'lennys-first-article-1-29b5102ef9a84eb9bfbf60da42f63122',
      body: "Like I said... I don't know what's going on. I'm just testing Hoslack's work.",
      description: "This is my first article and I don't know what's going on...",
      images: ['https://url-to-image'],
      createdAt: '2018-09-12T18:51:24.099861Z',
      updatedAt: '2018-09-12T18:51:24.100977Z',
      tagList: ['Lenny', 'First'],
    },
  ];
  it('Should render a loader if isFetching is true', () => {
    const tree = shallow(
      <MemoryRouter initialEntries={[{ pathname: '/', key: 'testKey' }]}>
        <ArticlesSection {...testProps} isFetching />
      </MemoryRouter>,
    );
    expect(shallowToJSON(tree)).toMatchSnapshot();
  });
  it('Should render the error message in case of failure', () => {
    const tree = shallow(
      <MemoryRouter initialEntries={[{ pathname: '/', key: 'testKey' }]}>
        <ArticlesSection {...testProps} fetchFailure errorMessage="An error occurred" />
      </MemoryRouter>,
    );
    expect(shallowToJSON(tree)).toMatchSnapshot();
  });

  it('Should render all articles in case of a successful request', () => {
    const tree = shallow(
      <MemoryRouter initialEntries={[{ pathname: '/', key: 'testKey' }]}>
        <ArticlesSection {...testProps} articles={articles} />
      </MemoryRouter>,
    );
    expect(shallowToJSON(tree)).toMatchSnapshot();
  });
});

describe('The Dashboard component', () => {
  it('Should not regress', () => {
    const tree = shallow(
      <MemoryRouter initialEntries={[{ pathname: '/', key: 'testKey' }]}>
        <Dashboard />
      </MemoryRouter>,
    );
    expect(shallowToJSON(tree)).toMatchSnapshot();
  });
});
