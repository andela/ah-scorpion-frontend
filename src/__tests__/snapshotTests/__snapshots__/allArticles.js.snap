// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Article component Should not regress 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "testKey",
        "pathname": "/",
        "search": "",
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Article
    author="lenny"
    averageRating={1}
    createdAt="2018-09-12T18:51:24.099861Z"
    description="test description"
    dislikes={0}
    image=""
    likes={0}
    slug="test-slug-001"
    title="test title"
  />
</Router>
`;

exports[`The ArticleList component Should render all articles in case of a successful request 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "testKey",
        "pathname": "/",
        "search": "",
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <ArticlesList
    articles={
      Array [
        Object {
          "author": Object {
            "bio": "",
            "email": "lennykmutua@gmail.com",
            "image": "",
            "username": "lenny",
          },
          "averageRating": null,
          "body": "Like I said... I don't know what's going on. I'm just testing Hoslack's work.",
          "createdAt": "2018-09-12T18:51:24.099861Z",
          "description": "This is my first article and I don't know what's going on...",
          "dislikes": 0,
          "favorited": 1,
          "id": 43,
          "images": Array [
            "https://url-to-image",
          ],
          "likes": 0,
          "ratingsCount": 0,
          "slug": "lennys-first-article-1-29b5102ef9a84eb9bfbf60da42f63122",
          "tagList": Array [
            "Lenny",
            "First",
          ],
          "title": "Lenny's First Article 1",
          "updatedAt": "2018-09-12T18:51:24.100977Z",
        },
      ]
    }
    errorMessage=""
    fetchFailure={false}
  />
</Router>
`;

exports[`The ArticleList component Should render text if there are no articles 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "testKey",
        "pathname": "/",
        "search": "",
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <ArticlesList
    articles={Array []}
    errorMessage=""
    fetchFailure={false}
  />
</Router>
`;

exports[`The ArticleList component Should render the error message in case of failure 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "testKey",
        "pathname": "/",
        "search": "",
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <ArticlesList
    articles={Array []}
    errorMessage="An error occurred"
    fetchFailure={true}
  />
</Router>
`;

exports[`The ArticleSection component Should render a loader if isFetching is true 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "testKey",
        "pathname": "/",
        "search": "",
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Connect(ArticlesSection)
    articles={Array []}
    errorMessage=""
    fetchFailure={false}
    isFetching={true}
  />
</Router>
`;

exports[`The ArticleSection component Should render all articles in case of a successful request 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "testKey",
        "pathname": "/",
        "search": "",
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Connect(ArticlesSection)
    articles={
      Array [
        Object {
          "author": Object {
            "bio": "",
            "email": "lennykmutua@gmail.com",
            "image": "",
            "username": "lenny",
          },
          "averageRating": null,
          "body": "Like I said... I don't know what's going on. I'm just testing Hoslack's work.",
          "createdAt": "2018-09-12T18:51:24.099861Z",
          "description": "This is my first article and I don't know what's going on...",
          "dislikes": 0,
          "favorited": 1,
          "id": 43,
          "images": Array [
            "https://url-to-image",
          ],
          "likes": 0,
          "ratingsCount": 0,
          "slug": "lennys-first-article-1-29b5102ef9a84eb9bfbf60da42f63122",
          "tagList": Array [
            "Lenny",
            "First",
          ],
          "title": "Lenny's First Article 1",
          "updatedAt": "2018-09-12T18:51:24.100977Z",
        },
      ]
    }
    errorMessage=""
    fetchFailure={false}
    isFetching={false}
  />
</Router>
`;

exports[`The ArticleSection component Should render the error message in case of failure 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "testKey",
        "pathname": "/",
        "search": "",
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Connect(ArticlesSection)
    articles={Array []}
    errorMessage="An error occurred"
    fetchFailure={true}
    isFetching={false}
  />
</Router>
`;

exports[`The Dashboard component Should not regress 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "testKey",
        "pathname": "/",
        "search": "",
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Dashboard />
</Router>
`;
